cluster-autoscaler/metrics/liveness.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type HealthCheck struct {
+	lastActivity time.Time
+	mutex        *sync.Mutex
+}
+
+const (
+	timeout = time.Minute * 10
+)
+
+var (
+	healthCheckInstance = &HealthCheck{
+		lastActivity: time.Now(),
+		mutex:        &sync.Mutex{},
+	}
+)
+
+func (hc *HealthCheck) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if time.Now().After(hc.lastActivity.Add(timeout)) {
+		w.WriteHeader(500)
+		w.Write([]byte(fmt.Sprintf("Error: last activity more than %v ago", time.Now().Sub(hc.lastActivity).String())))
+	} else {
+		w.WriteHeader(200)
+		w.Write([]byte("OK"))
+	}
+}
+
+func HealthCheckHandler() http.Handler {
+	return healthCheckInstance
+}
+
+func UpdateLastActivity(timestamp time.Time) {
+	if timestamp.After(healthCheckInstance.lastActivity) {
+		healthCheckInstance.mutex.Lock()
```

Comment:
You cannot read lastActivity variable outside of a mutex. Move it above line 57.